pkg/assets/builder.go

```diff
@@ -127,32 +148,117 @@ func (a *AssetBuilder) RemapImage(image string) (string, error) {
 	}
 
 	a.ContainerAssets = append(a.ContainerAssets, asset)
-
 	return image, nil
 }
 
 // RemapFile sets a new url location for the file, if a AssetsLocation is defined.
-func (a AssetBuilder) RemapFile(file string) (string, error) {
-	if file == "" {
-		return "", fmt.Errorf("unable to remap an empty string")
+func (a *AssetBuilder) RemapFileAndSHA(file *url.URL, sha *url.URL) (*url.URL, *hashing.Hash, error) {
+	if file == nil {
+		return nil, nil, fmt.Errorf("unable to remap an empty string")
 	}
 
 	fileAsset := &FileAsset{
-		File:              file,
-		CanonicalLocation: file,
+		FileURL:    file,
+		SHAFileURL: sha,
 	}
 
 	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
-		fileURL, err := url.Parse(file)
+		fileAsset.CanonicalFileURLLocation = file
+		fileAsset.CanonicalSHAFileURLLocation = sha
+
+		file, err := a.normalizeURL(file)
 		if err != nil {
-			return "", fmt.Errorf("unable to parse file url %q: %v", file, err)
+			return nil, nil, fmt.Errorf("unable to parse file url %q: %v", file, err)
 		}
 
-		fileRepo := strings.TrimSuffix(*a.AssetsLocation.FileRepository, "/")
-		fileAsset.File = fileRepo + fileURL.Path
+		fileAsset.FileURL = file
+
+		sha, err = a.normalizeURL(sha)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to parse sha url %q: %v", file, err)
+		}
+
+		fileAsset.SHAFileURL = sha
+		glog.V(4).Infof("adding remapped file: %+v", fileAsset)
+	}
+
+	h, err := a.FindHash(fileAsset)
+	if err != nil {
+		return nil, nil, err
 	}
 
+	fileAsset.SHAValue = h.Hex()
 	a.FileAssets = append(a.FileAssets, fileAsset)
 
-	return fileAsset.File, nil
+	glog.V(8).Infof("adding file: %+v", fileAsset)
+
+	return fileAsset.FileURL, h, nil
+}
+
+// RemapFileAndSHAValue is used exclusively to remap the cni tarball, as the tarball
+// does not have a sha file in object storage.
+func (a *AssetBuilder) RemapFileAndSHAValue(file *url.URL, shaValue string) (*url.URL, error) {
+	if file == nil {
+		return nil, fmt.Errorf("unable to remap an empty string")
+	}
+
+	fileAsset := &FileAsset{
+		FileURL:  file,
+		SHAValue: shaValue,
+	}
+
+	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
+		fileAsset.CanonicalFileURLLocation = file
+
+		file, err := a.normalizeURL(file)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse file url %q: %v", file, err)
+		}
+
+		fileAsset.FileURL = file
+		glog.V(4).Infof("adding remapped file: %q", fileAsset.FileURL.String())
+	}
+
+	a.FileAssets = append(a.FileAssets, fileAsset)
+
+	return fileAsset.FileURL, nil
+}
+
+// FindHash returns the hash value from remove sha file via https.
+func (a *AssetBuilder) FindHash(file *FileAsset) (*hashing.Hash, error) {
```

Comment:
This function should be private, because otherwise it is very surprising that it does not consider SHAFileURL & friends